Added contributing policy

dbrgn opened the pull request:
Fixes #14.

dbrgn committed:
Added contributing policy

Fixes #14.

doismellburning commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,35 @@
+contribution guidelines though:
+
+- Run the [test suite](https://github.com/py3minepi/py3minepi#testing) before
+  committing and pushing your changes.
```
I don't quite feel comfortable asking people to do things I wouldn't myself - we have Travis running PR tests after all...

dbrgn replied:
It already avoids a lot of messy pull requests (e.g. non-PEP8) though. A lot of people simply ignore failing Travis tests, so that's an explicit hint that the tests are important :) Some also don't run the tests because they don't know how to, so the link in here points them to the section in the README.

ghickman replied:
I'd be pro keeping this in but maybe turning it into a nice-to-have. As @dbrgn points out it's likely to save messier PRs but we shouldn't enforce it given we do have Travis catching any errors.

dbrgn replied:
I'll move it to the second section, OK?

ghickman replied:
Sounds good to me.

doismellburning replied:
People will learn to not ignore Travis failures when PRs that don't pass don't get merged ;) - point taken though!

doismellburning commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,35 @@
+  relevant).  [This](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
+  is a nice blog post concerning git commit messages.
+- Add yourself to the [AUTHORS.txt
+  file](https://github.com/py3minepi/py3minepi/blob/master/AUTHORS.txt).
```
Nice; what do you think about dropping the link though? (because it's fixed branch-wise)

dbrgn replied:
We could drop it, but pointing to master isn't a problem as that's the currently valid file :) If you think it's better I'll remove it though.

doismellburning replied:
Good point!

ghickman commented:
@dbrgn – Great work!

Could you replace the part about `hub` turning an issue into a pull request? I'd love to be able to suggest people use this but after using it myself tonight it gives a deprecation warning. :disappointed:

doismellburning commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,35 @@
+The following items make our lifes easier, but are optional:
+
+- If you know how to use `git rebase`, please rebase/sqash your commits so that
+  unnecessary noise in the commit history is avoided.
```
I absolutely disagree with this...

dbrgn replied:
I'm not saying that everything should be in 1 commit. I'm saying that the following commits:
- attempt at fixing issue X
- typo
- missed import

...should become...
- Fixed issue X

Do you agree? In that case I'll improve the wording.

doismellburning replied:
No, I still disagree... Do others have opinions?

dbrgn replied:
For what reasons? It's basically free (if you know how to deal with interactive rebase). And the biggest advantage is that 1 feature can be traced back to 1 commit, without being split up into multiple commits.

E.g. in this case I'd squash all commits into a single one that introduces the contributors agreement that we've all agreed on.

doismellburning replied:
It rewrites history, it hides the reality, and it interferes with people collaborating on the same PR.

One feature can be traced back to one commit anyway: the merge commit

dbrgn replied:
> It rewrites history

But only pull request history. It has not become project history yet, therefore it is not important. I agree that you should never ever rebase the master branch. But nobody should check out a pull request branch and start to use it productively.

Many commit messages that made sense in the context of a pull request don't make sense in the context of the full repository history anymore.

It also prevents that a commit for a feature contains great documentation on the reasons for this commit. While that reasoning might be present in a pull request discussion, that won't be obvious to a user or dev later on. If it's in the git history, it can be looked up.

> and it interferes with people collaborating on the same PR.

Usually only one person should contribute per pull request (unless the people are in the same room and know they don't cause conflicts).

> One feature can be traced back to one commit anyway: the merge commit

A merge commit doesn't show the entire diff.

doismellburning replied:
> Many commit messages that made sense in the context of a pull request don't make sense in the context of the full repository history anymore.

Correct, but they make sense in the context of their parent and in context of the branch

> It also prevents that a commit for a feature contains great documentation on the reasons for this commit. While that reasoning might be present in a pull request discussion, that won't be obvious to a user or dev later on. If it's in the git history, it can be looked up.

That completely differs with my experience; it is not prevented at all

> Usually only one person should contribute per pull request (unless the people are in the same room and know they don't cause conflicts).

Fair enough; I've worked in a lot of situations where it's been considered totally reasonable to fix obvious mistakes in someone else's PR, but can see that not working in all scenarios

> A merge commit doesn't show the entire diff.

You may be looking for `-m` (alas I don't think it has a descriptive longopt)

dbrgn replied:
> Correct, but they make sense in the context of their parent and in context of the branch

Which in turn is hard to see, you need graphical visualization tools just to be able to find the historic branches. Which gets _really_ messy, especially with multiple merges from master into the branch.

Btw, I'm not advocating rebasing the feature into master. You still have merge commits, even if you squash some of your commits in the PR.

> That completely differs with my experience; it is not prevented at all

It is not prevented, but you need to find the commit with the good documentation among multiple fix- and typo-commits which don't add any "real" value by themselves.

> You may be looking for -m (alas I don't think it has a descriptive longopt)

Github doesn't show it though. It's unnecessary additional work.

As a realworld example, see the following three commits:

https://github.com/celery/celery/commit/133f2a1aecf5c6bb6e7e04978d14d041a9706c42
https://github.com/celery/celery/commit/49d356d61a67bed94ef343ae588928873ca35867
https://github.com/celery/celery/commit/05e2567ede7c71f523cc5783ccb47ad8eac951d8

While the commits are fine, you need to jump around between all three of them just to get "the big picture". One of the commits moves around one of the new methods (which appears as a deletion and an addition) and adds a docstring to a method that was created in one of the previous two commits. While this is not "bad", it would be much nicer if there were one commit adding the method including a good docstring. While the second commit could stay separate as it adds value (-> a new method), the third one does not add any value on its own.

ghickman replied:
### Well Known Projects Maintaining Clean History

Django and Celery are great examples of this but for every project doing this there are major projects not doing it. Rebasing and squashing for this purpose has, to my mind anyway, always appeared to be a very philosophical project specific decision. Personally I feel it raises the barrier of entry to a project, which I'm keen to avoid. Taking Django as an example, squashing is typically done by the merging developer, not the person raising the PR, to avoid the raiser having to know how to do it. This puts the burden of knowing how to squash and rebase on the core dev team. Now, I'm more than capable of doing some fairly advanced things with git's history manipulation tools but I don't think this should be a requirement of any of the other developers, nor do I want to have to get involved with PRs myself.

### It's an Advanced Feature

Rebasing is, at the end of the day, one of the most advanced features in Git. Adding such a powerful and dangerous tool as a suggested method for contribution (even an optional one) is likely to bite us in the long run.

### Maintaining History

Personally I dislike the use of rebase once code has been pushed, it's far too easy to impact other developers (whether you mean to or not) and rewriting history can have nasty knock on effects. Take this PR: if you rebase the history now where do all our comments go? IIRC they disappear because they're tied to the SHAs of your commits. I'd much rather have some messy commits which only make contextual sense in the history _and_ the PRs on GitHub than clean commits in the history.

As @doismellburning has mentioned there are tools for viewing even messy commits with context on the CLI and every GUI tool I know of can do this.

### TL;DR

I love using rebase and having a clean history but have been bitten so many times by pushing rebased commits to collaborative projects that I cannot in any way condone adding this to the contributing docs.

dbrgn replied:
> Take this PR: if you rebase the history now where do all our comments go? IIRC they disappear because they're tied to the SHAs of your commits.
> No, they stay because you commented on the diff, not on the commits themselves.

I'm really a fan of clean commit histories. Also, the policy clearly states `If you know how to use "git rebase"`, so nobody will be forced to do anything he/she isn't capable of doing.

But if you both agree, I'll remove that section.

pozorvlak replied:
Damn it, looks like I'm too late. I really really _really_ like the use of git-rebase to produce clean and easy-to-review histories; I use it for that at work every day. The commit history should, IMHO, tell a _useful_ story about how the code was developed; this is not always the same thing as a _literally correct_ history. The comments are not lost by GitHub: they show up as "commented on an outdated diff" links. But since most of the uses of this feature are to address minor issues, you don't usually need to read them anyway.

On [http://github.com/Growstuff/growstuff](Growstuff), "reorder your commits if you know how" is stated policy, but I think I'm the only one who actually does. So I don't see any problem with having an "if you know how" commit-rewriting policy.

doismellburning commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,35 @@
+  and then issue the following command: `hub pull-request -b py3minepi:master -i
+  <issue-number> -h <your-github-username>:<your-branch-name>` This is no strict
+  requirement though, if you don't have hub installed or prefer to use the web
+  interface, then feel free to post a traditional pull request.
```
Nifty, but I think this is deprecated? /cc @ghickman

doismellburning commented on README.rst:
```diff
@@ -31,6 +32,61 @@ Goals
+Testing
+-------
+
+Testing for swidGenerator is set up using `Tox <http://tox.readthedocs.org/>`_
```
swidGenerator?

dbrgn replied:
Aaaah. Copy pasta.

dbrgn commented:
> Could you replace the part about hub turning an issue into a pull request? I'd love to be able to suggest people use this but after using it myself tonight it gives a deprecation warning. :disappointed:

There's an interesting discussion here: https://github.com/github/hub/commit/4f70dd126f46dec14fc341c97c18efae417743c7 Looks like it might actually be removed. That would be a bummer though.

Do you think we should stop recommending it even though it _might_ not work anymore _in the future_? We could also benefit from it right now and remove the line if it actually doesn't work anymore.

(I could also agree to removing it though.)

doismellburning commented:
Regarding hub, although it's neat, it's not _that_ awesome, thus I'd rather not start encouraging it 'til it's definitely not going away

doismellburning commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,27 @@
+- Run the [test suite](https://github.com/py3minepi/py3minepi#testing) before
+  committing and pushing your changes.
+- If you know how to use `git rebase`, please rebase/sqash your commits so that
+  unnecessary noise in the commit history is avoided.
```
See https://github.com/py3minepi/py3minepi/pull/31#discussion_r18306501

ghickman commented:
@dbrgn – Thanks!

dbrgn commented:
Removed hub. The issue regarding the rebasing of pull requests (https://github.com/py3minepi/py3minepi/pull/31#discussion-diff-18306501) is still open though. I'd appreciate comments from other developers though (@ghickman, @bennuttall etc). Keep in mind that this "request" is in the "nice to have, but not necessarily needed" section.

dbrgn commented:
I removed the rebasing suggestion (see other comment). I hope you don't mind that I squashed my history in this case - I used some sloppy, non-descriptive commit messages ;)

In the future I'll try to write better (in global context) commit messages even in branches.

PS: If a pull request got a bit behind master and doesn't merge cleanly anymore, would you also not suggest rebasing against master?

doismellburning commented:
Cheers - I would indeed object to rebasing against master on similar grounds: why throw away history and the original context in which the branch was made

dbrgn commented:
As long as you don't squash, you don't loose history because a rebase keeps track of both the commit date and the original authoring date (when was the commit originally created vs when was this actuall commit id generated).

The original context is also not very important, because if the context changed significantly then the PR needs to be changed or thrown away. When merging, only the current context matters :)

Anyways, this PR is ready to merge now.

pozorvlak commented:
The answer to "why throw away history" is "because it's irrelevant and obscures the bits of history that reviewers actually want to see". If your branch contains a bunch of "merge master into $feature" commits then the overall diff (as shown by GitHub) becomes bloated with a bunch of stuff that has no connection to the feature that the PR implements.

dbrgn commented:
Thanks :) This PR is not merged yet, so it's not too late to discuss it.

dbrgn committed:
Added style guide note about future imports

Refs #45.

dbrgn commented:
So, any consent? Can we maybe keep it an optional thing? I find that @pozorvlak provides a few very compelling reasons. And as long as it's in a pull request only, it does not hurt anyone.

doismellburning commented:
While I still disagree with any encouragement of rewriting/hiding/obscuring history, we actually _do_ agree on the current content of this PR, so merging it in!
